### ISSUE TITLE
Incomplete package update

### ISSUE THREAD

**Issue by Djaler** [pnpm/pnpm#2222]
### Code to reproduce the issue:

I created a repo that reproducing the issue https://github.com/Djaler/pnpm-incomplete-update-bug.

1. Initial state. Install `@stryker-mutator/core@2.4.0` and `typescript@3.7.2` (https://github.com/Djaler/pnpm-incomplete-update-bug/commit/6fc2ffb05934ce95698db679d9a0b359616bd9a8)
2. Change `typescript` version to `3.7.3` (https://github.com/Djaler/pnpm-incomplete-update-bug/commit/b1182b9fc947917846c65ce5df6f69355b793509)
3. Install `left-pad` (https://github.com/Djaler/pnpm-incomplete-update-bug/commit/a861cab0512d2e3ed5ba572d63e7168f64a80c5f)

### Expected behavior:

At 2 step typescript 3.7.2 should gone

### Actual behavior:

At 2 step `typescript@3.7.2` still exists in dependencies of `typed-inject`. But, on 3 step it was replaced by `typescript@3.7.3` completely

### Additional information:

 - `node -v` prints: v10.17.0
 - Windows, OS X, or Linux?: Linux

**Comment by zkochan** [comment]
Related #1041

**Comment by Djaler** [comment]
Note: I don' use named install. I change version in package.json and then run simple `pnpm i`

**Comment by Djaler** [comment]
Also, the main strangeness for me is a things that happened on 3 step

**Comment by Djaler** [comment]
@zkochan sorry for triggering you, but I really interested in solving this problem. Do you have any plans for it?

**Comment by zkochan** [comment]
yes, but it will require a complex, big, and dangerous refactoring.

**Comment by Djaler** [comment]
Update on pnpm 5.1.1: now old version not deleted at all

**Comment by Djaler** [comment]
Maybe something else can help in this case? Like npm dedupe

**Comment by zkochan** [comment]
It can be done but the resolution algorithm of pnpm is [very complex code](https://github.com/pnpm/pnpm/blob/master/packages/resolve-dependencies/src/resolveDependencies.ts) optimized for speed. Any change may cause a huge performance regression. While developing v5 I have caused a few regressions and it was hard to track the issues. That code needs some refactoring and more granular test coverage.

Here's one issue I created for that: https://github.com/pnpm/pnpm/issues/2548

Also, I think it might make sense to create a separate resolution algorithm for a clean install, when there is no lockfile present. Maybe even 2 new algorithms. The complexity of that file is caused due to the many different scenarios it should handle. No lockfile, up-to-date lockfile, update, etc.

**Comment by Djaler** [comment]
Is there any way I can manually dedup packages after install?

**Comment by zkochan** [comment]
Doing a `pnpm update --depth Infinity` will dedupe dependencies.

**Comment by Djaler** [comment]
This will update all packages to the latest versions. But what if I want to update only one dependency to the exact version?

**Comment by zkochan** [comment]
it is currently not possible. If you need to pin a dependency to some specific version, you can use a hook in a pnpmfile.js

**Comment by zkochan** [comment]
We can update `pnpm update foo --depth Infinity` so that it updates `foo` on all levels. Currently, `pnpm update foo` would only update foo in direct dependencies.

**Comment by Djaler** [comment]
> pnpm update foo --depth Infinity

Didn't work for me

**Comment by zkochan** [comment]
I know, because it was not implemented yet

**Comment by Djaler** [comment]
> I know, because it was not implemented yet

Ah, I misunderstood you.

**Comment by zkochan** [comment]
@Djaler with pnpm v5.4.5, no `pnpm update foo --depth Infinity` will work as expected.

**Comment by Djaler** [comment]
If someone here uses renovatebot too, you can run `pnpm update foo --depth Infinity` automatically after install:
https://github.com/renovatebot/renovate/pull/7632

**Comment by Djaler** [comment]
Hmm, pnpm v5.11.1 updates all dependencies, not just `foo`

**Comment by zkochan** [comment]
It is hard to fix all the different use cases in one go. I think I will try to fix one particular issue next:

In a workspace, when two projects have the same dependency with the same version range, the dependency should be always resolved to the same version in both projects.

**Comment by attilapuskas** [comment]
A potential workaround seemed to work for me to add a temporary entry to the [pnpm.overrides](https://pnpm.io/package_json#pnpmoverrides).
In case of the original example it can be:
`
"pnpm": {
    "overrides": {
      "typescript": "3.7.3"
    }
}
`
